### ISSUE TITLE
Extremely slow eigendecomposition compared to numpy/scipy.

### ISSUE THREAD

**Issue by seanslice** [tensorflow/tensorflow#42967]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): macOS 10.15.6
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.3.0-rc2-23-gb36436b087 2.3.0
- Python version: 3.7.7
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

I am using eigendecomposition in Tensorflow and find that it is extremely slow. This is on a mac, so these are CPU computations. But I've also done this on a linux box with a GPU and found the same thing. Here's the code to show Tensorflow's speed vs numpy and scipy:

```
import numpy as np
import scipy as sp
import tensorflow as tf
from time import time

A = np.random.randn(400, 400)
A_tf = tf.constant(A)

cur = time()
d, v = sp.linalg.eig(A)
print(f'sp: {time() - cur:4.2f} s')

cur = time()
d, v = np.linalg.eig(A)
print(f'np: {time() - cur:4.2f} s')

cur = time()
d, v = tf.linalg.eig(A_tf)
print(f'tf: {time() - cur:4.2f} s')
```
This gives the following output:
```
sp: 0.09 s
np: 0.08 s
tf: 5.04 s
```
Any ideas of what's up here?

**Comment by bhack** [comment]
Ref. Stackoverflow https://stackoverflow.com/questions/63746426/tensorflow-eigenvalue-decomposition-is-extremely-slow

**Comment by seanslice** [comment]
I posted both. Wasn't sure which venue was the correct one.

**Comment by bhack** [comment]
Internally it is just a call to the eigen library: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/linalg/eig_op_impl.h#L77

/cc @randl

**Comment by bhack** [comment]
Also as we are using  `Eigen::ComplexEigenSolver` this is generally classified as a quite slow solution: [Check this table](https://eigen.tuxfamily.org/dox/group__TopicLinearAlgebraDecompositions.html).  
> As it is interactive their convergence speed depends on how well the eigenvalues are separated.

So probably `np.random.randn(400, 400)` it is not one of the best input case for this API call in Eigen.

**Comment by bm777** [comment]
Depending on your query, I used TensorFlow 1.5 on ubuntu 18.04.
I used TensorfFlow.linalg.eigh instead. it was fast that two others.
```
import numpy as np
import scipy as sp
import tensorflow.compat.v2 as tf
from time import time

A = np.random.randn(400, 400)
A_tf = tf.constant(A)

cur = time()
d, v = sp.linalg.eig(A)
print(f'sp: {time() - cur:4.2f} s')

cur = time()
d, v = np.linalg.eig(A)
print(f'np: {time() - cur:4.2f} s')

cur = time()
d, v = tf.linalg.eigh(A_tf)    # in tensofow 1.5
print(f'tf: {time() - cur:4.2f} s')


```

The result.
![result](https://user-images.githubusercontent.com/29865600/92306034-2e50f300-ef84-11ea-8fd4-cc89e9ff6000.png)

**Comment by seanslice** [comment]
Any reason to not use a faster solver? Or is there a hack to switch to a different one? My use case is an optimization that includes a sum of a random NxN matrix plus a parameterized low-rank component.

**Comment by seanslice** [comment]
tf.linalg.eigh is only for symmetric matrices. It didn't give you an error because it ignores the upper triangle of the input.

**Comment by seanslice** [comment]
@bhack Things are a bit more equal when using a complex matrix as input to the numpy and scipy solvers, but there is still a very significant difference which renders my optimization unworkable.
```
import numpy as np
import scipy as sp
import tensorflow as tf
from time import time

A = np.random.randn(400, 400).astype(complex)
A_tf = tf.constant(A)

cur = time()
d, v = sp.linalg.eig(A)
print(f'sp: {time() - cur:4.2f} s')

cur = time()
d, v = np.linalg.eig(A)
print(f'np: {time() - cur:4.2f} s')

cur = time()
d, v = tf.linalg.eig(A_tf)
print(f'tf: {time() - cur:4.2f} s')
```
With output:
```
sp: 0.22 s
np: 0.17 s
tf: 4.53 s
```

**Comment by bhack** [comment]
Can you try to code and benchmark a minimal c++ example for your use case with [Spectra](https://spirit-docs.readthedocs.io/en/latest/core/thirdparty/spectra/README.html)?

**Comment by bhack** [comment]
@seanslice Also you will probably find similar `np.linalg.eig` performance compiling Eigen with `EIGEN_USE_LAPACKE`. See https://eigen.tuxfamily.org/dox-devel/TopicUsingBlasLapack.html`

**Comment by seanslice** [comment]
@bhack I need all the eigenvalue and vectors, so Spectra (which returns a subset of the values/vectors) isn't the right solution. I'm varying matrix sizes between 50 and 1000, which is well within the range of relatively rapid eigendecomposition on modern hardware.

My use case is below. The loss function is a bit of a mess of linear algebra operations, but the limiting step is the call to eig() during the forward pass.

```
import tensorflow as tf
import numpy as np
from time import time

# meta-parameters
N = 100
n = 40
epochs = 1001
print_interval = 200
beta = 0.05
ext_penalty = 1e5
np.random.seed(0)

# fixed values
J0 = np.random.randn(N, N) / np.sqrt(N)
w = np.random.randn(N) / np.sqrt(N)

eig_time = [0.0]
# loss function; store time performing eig()
def loss_f(U, V):
    U = U / tf.linalg.norm(U, axis=-2, keepdims=True)
    V = V / tf.linalg.norm(V, axis=-1, keepdims=True)
    J = J0 + U @ V
    pre_eig = time()
    d, r = tf.linalg.eig(J)
    eig_time[0] += time() - pre_eig
    L = tf.linalg.inv(tf.linalg.adjoint(r) @ r)
    g = d - 1
    g_real = tf.math.real(g)
    tmp = -g[:, None] - tf.math.conj(g)
    G1 = tmp / (tmp * tmp + 1e-20)
    G2 = (g[:, None] * tf.math.conj(g)) * G1
    wR = tf.linalg.matvec(r, tf.cast(w, tf.complex128), transpose_a=True)
    return tf.math.real(tf.reduce_sum((r @ (L * G1)) * tf.math.conj(r)) / N + beta * tf.reduce_sum(wR * tf.linalg.matvec(L * G2, tf.math.conj(wR)))) \
           + ext_penalty * tf.reduce_sum(tf.where(g_real < 0, tf.zeros_like(g_real), g_real)**2)

# weights to be optimized
U = tf.Variable(np.random.randn(N, n) / np.sqrt(N))
V = tf.Variable(np.random.randn(n, N) / np.sqrt(N))

# convenience fx for printing losses & execution time during optimization
def stager():
    loss = loss_f(U, V)
    if stager.count % print_interval == 0:
        print(f'iter {stager.count:4d}, time {time() - stager.time:5.2f}s, eig_time {eig_time[0]:5.2f}s, loss {loss:7.5f}')
        stager.time = time()
        eig_time[0] = 0.0
    stager.count += 1
    return loss
stager.count = 0
stager.time = time()

# optimization routine
opt = tf.keras.optimizers.Adam(learning_rate=0.01)
for _ in range(epochs):
    opt.minimize(stager, [U, V])
```
Here's sample output for a few steps of optimization for a matrix of 100x100 (`N=100`, `n=40`, `epochs=1001`, `print_interval=200`):
```
iter    0, time  0.09s, eig_time  0.09s, loss 4656.80738
iter  200, time 17.48s, eig_time 14.64s, loss 2.00800
iter  400, time 17.69s, eig_time 14.82s, loss 1.72453
iter  600, time 17.67s, eig_time 14.81s, loss 1.57498
iter  800, time 17.59s, eig_time 14.76s, loss 1.47659
iter 1000, time 17.76s, eig_time 14.90s, loss 1.40378
```
And here's for a matrix of 400x400 (`N=400`, `n=160`, `epochs=21`, `print_interval=4`):
```
iter    0, time  4.52s, eig_time  4.45s, loss 11191.68619
iter    4, time 20.34s, eig_time 19.52s, loss 12.15866
iter    8, time 20.34s, eig_time 19.50s, loss 4.54211
iter   12, time 19.64s, eig_time 18.80s, loss 3.22969
iter   16, time 19.61s, eig_time 18.78s, loss 2.80901
iter   20, time 19.05s, eig_time 18.24s, loss 2.61946
```

**Comment by bhack** [comment]
If you need all the values and something more like `np.linalg.eig` you can call LAPACK `?gees` but you need to recompile Tensorflow from source and modify a little bit the bazel flags for Eigen to enable `EIGEN_USE_LAPACKE`.  

If you are adventurous you can copy some build tips from: 
https://gitlab.com/arm-hpc/packages/-/wikis/packages/tensorflow?version_id=5c32ca4f0132af89f79db6303437d6e66526364b#alter-eigens-defines-includes-and-depencencies

After you have patched the build rules you could build TF from source as usual:
https://www.tensorflow.org/install/source

**Comment by bhack** [comment]
P.s. In the meantime waiting for other maintainers comments if they want to suggest you an alternative formulation of your loss.

**Comment by seanslice** [comment]
Thanks a lot for your input. I really appreciate it. I haven't tried to compiling from source. I admit to being a bit daunted by the prospect, but I'll dive in and see how it goes. Thanks again!

**Comment by bhack** [comment]
@seanslice As building Tensorflow it is quite a resource intensive and long process I suggest you to directly prototype standalone  c++ code like:

```
#include <iostream>
#include <Eigen/Eigenvalues>
#include <Eigen/Dense>
#include <chrono>

using namespace std;
using namespace Eigen;

int main( int argc, const char* argv[] )
{
	auto start = chrono::steady_clock::now();
	MatrixXcf A = MatrixXcf::Random(400,400); 
	ComplexEigenSolver<MatrixXcf> ces;
	ces.compute(A);
	auto end = chrono::steady_clock::now();
        auto diff = end - start;
        cout << chrono::duration <double, milli> (diff).count() << " ms" << endl;
}
```

`g++  -I /usr/include/eigen3 eig_bench.cc -o eig_bench -llapack -llapacke -lblas  -D EIGEN_USE_LAPACKE=1 -lm`

**Comment by seanslice** [comment]
It took me awhile to get openblas and lapack properly installed and discoverable by the compiler. In addition to the C++ benchmark code you wrote, I made to little python scripts. `eig_test_np.py`:
```
import numpy as np
from time import time

J = np.random.randn(400, 400)
cur = time()
d, v = np.linalg.eig(J)
print(f'{1000 * (time() - cur):.2f} ms')
```
And `eig_test_tf.py`:
```
import tensorflow as tf
from time import time

J = tf.random.normal((400, 400), dtype=tf.float64)
cur = time()
d, v = tf.linalg.eig(J)
print(f'{1000 * (time() - cur):.2f} ms')
```
It's pretty clear from the output that numpy is not using `ComplexEigenSolver`:
```
$ ./eig_bench
2626.33 ms
$ python eig_test_tf.py
4817.74 ms
$ python eig_test_np.py
106.44 ms
```

**Comment by bhack** [comment]
I suppose you need to test with something like:
```
a = np.random.randn(400,400)+np.random.randn(400,400)*1j
w, v = np.linalg.eig(a)
```

As you can see from numpy doc:
https://numpy.org/doc/stable/reference/generated/numpy.linalg.eig.html#numpy-linalg-eig
> This is implemented using the _geev LAPACK routines

That with the c++ compiler flag theorically  could be the same of Eigen as you can see in the linked table:
https://eigen.tuxfamily.org/dox-devel/TopicUsingBlasLapack.html

**Comment by refraction-ray** [comment]
This is ultimately due to single threaded Eigen implementations for eig op which could be linked to multithreaded MKL but actually not in tf bazel setup. There are many similar problems complaining the speed as  https://github.com/tensorflow/tensorflow/issues/7128, https://github.com/tensorflow/tensorflow/issues/13222, etc., and this problem can only be fully addressed and solved by https://github.com/tensorflow/tensorflow/issues/34924. Namely, by supporting MKL linkage for eigen when compiling tensorflow, but as described in the above issue, I am not fully clear how to make such setup work.

For now there is a workaround using `tf.py_function` though, which can utilize ``eig`` from numpy or scipy in forward pass (which is ultimately provided by multithreded MKL or openblas) and enjoy automatic differentiation at the same time. See a full demo below (the gradient code part is directly copied from tf codebase):

```python
def _SafeReciprocal(x, epsilon=1E-20):
      return x / (x * x + epsilon)

@tf.custom_gradient
def my_eig(x):
    e, v = np.linalg.eig(x)
    def grad(grad_e, grad_v):
        vt = tf.linalg.adjoint(v)
        f = tf.linalg.set_diag(
          _SafeReciprocal(
              tf.expand_dims(e, -2) - tf.expand_dims(e, -1)),
          tf.zeros_like(e))
        f = tf.math.conj(f)
        vgv = tf.matmul(vt, grad_v)
        mid = tf.linalg.diag(grad_e)
        diag_grad_part = tf.linalg.diag(tf.linalg.diag_part(
          tf.cast(tf.math.real(vgv), vgv.dtype)))
        mid += f * (vgv - tf.matmul(tf.matmul(vt, v), diag_grad_part))
        grad_a = tf.linalg.solve(vt, tf.matmul(mid, vt))

        return tf.cast(grad_a, x.dtype)
    return (e, v), grad
    
a = tf.random.normal([400,400])
with tf.GradientTape() as tape:
    tape.watch(a)
    y = tf.py_function(func=my_eig, inp=[a], Tout=[tf.complex64, tf.complex64])
    print(y)
    l =  y[1][2,2]**2/tf.linalg.norm(y[1][:,2])
print(tape.gradient(l, a))
```

Such an approach provide speed similar to numpy eig and compatible with tf's AD infrastructure. The 400*400 case with gradient calculation requires time around 0.5s while a pure ``tf.eig`` foward pass with back propagation needs 8.5s.

Of course, this is not a perfect solution as ``py_function`` has many limitations such as it cannot be serialized, but I guess it is ok for most research cases when one just plays some small things with python.

**Comment by bhack** [comment]
I don't know If the table is still updated but I don't think also enabling MKL flag and linking has an effect on that call:

https://eigen.tuxfamily.org/dox/TopicUsingBlasLapack.html

I think the main difference with that flag in the c++ standalone and numpy is lapack `?gees` vs `?geev`.

Probably the standalone example could freely use multithread but I've not checked Tensorflow flags for the Eigen library.

[?gees doc](https://software.intel.com/content/www/us/en/develop/documentation/mkl-developer-reference-c/top/lapack-routines/lapack-least-squares-and-eigenvalue-problem-routines/lapack-least-squares-and-eigenvalue-problem-driver-routines/nonsymmetric-eigenvalue-problems-lapack-driver-routines/gees.html)

[?geev doc](https://software.intel.com/content/www/us/en/develop/documentation/mkl-developer-reference-c/top/lapack-routines/lapack-least-squares-and-eigenvalue-problem-routines/lapack-least-squares-and-eigenvalue-problem-driver-routines/nonsymmetric-eigenvalue-problems-lapack-driver-routines/geev.html)

**Comment by seanslice** [comment]
@refraction-ray This solution solved the problem. I achieved about a 20x speed up on my MacBook for `N=500`. With `tf.linalg.eig`, 25 iterations of ADAM took ~300 s. With `py_function` wrapping `np.linalg.eig`, 25 iterations took ~14 s.

Thank you tremendously for your help here!!

**Comment by ravikyram** [comment]
I was able to reproduce the issue with TF version 2.3, nightly version(`2.4.0-dev20200906`) and was able to reproduce the issue.Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/0d4767d0179310577ef82560db0e6f01/untitled310.ipynb#scrollTo=aJMXFa70L9OS).Thanks!

**Comment by dobos** [comment]
Also, `tf.linalg.eigh` uses way too much memory, this could be the reason why it's slow (when it fits into the GPU RAM, at least).

```
    def test_eigh_oom(self):
        cov = Uniform(0, 1, dtype=tf.float64).sample((5, 100000, 4, 4))
        cov += tf.transpose(cov, perm=[0, 1, 3, 2])

        s, u, v = tf.linalg.svd(cov, full_matrices=True)
        e, v = tf.linalg.eigh(cov)
```

SVD works, eigh gives OOM.

**Comment by dobos** [comment]
I just want to add that `tf.linalg.svd` without the `full_matrices=True` parameter also results in OOM if the input is large.
